corva/models/stream.py

```diff
@@ -28,24 +23,44 @@ class Record(BaseEventData):
     data: RecordData
 
 
-class AppMetadata(BaseEventData):
+class AppMetadata(CorvaBaseModel):
     app_connection_id: int
     app_version: Optional[int] = None
 
 
-class StreamEventMetadata(BaseEventData):
+class StreamEventMetadata(CorvaBaseModel):
     app_stream_id: int
+    source_type: Optional[str] = None
     apps: Dict[str, AppMetadata]
 
 
-class StreamEventData(BaseEventData):
+class StreamEventData(CorvaBaseModel):
     app_key: Optional[str] = None
     records: List[Record]
     metadata: StreamEventMetadata
+    asset_id: int = None
 
-    @property
-    def asset_id(self) -> int:
-        return self.records[0].asset_id
+    @pydantic.validator('asset_id', pre=True, always=True)
+    def set_asset_id(cls, v, values):
+        """dynamically sets value for asset_id
+
+        asset_id could've been defined as property like below.
+
+        @property
+        def asset_id(self) -> Optional[int]:
+            return self.records[0].asset_id if self.records else None
+
+        The issue with the above method is:
+         after filtering, we may end up with empty records. Which leads to asset_id becoming None.
+         Using this validator we are able to dynamically set and store value of asset_id,
+         no matter what happens to records.
+        """
+
+        records = values['records']  # type: List[Record]
+        if records:
+            return records[0].asset_id
+
+        raise ValueError('Can\'t determine asset_id as records are empty (which should not happen).')
```

Comment:
This should be double quotes, since you have a single quote in the sentence.

Reply:
the single quote in the sentence is escaped, so this works fine